File: merge_open_prs.py
#!/usr/bin/env python3
import subprocess
import json
import sys
import time

def run_cmd(cmd, check=True, capture=True):
    """Run a shell command"""
    try:
        if capture:
            result = subprocess.run(cmd, shell=True, check=check, 
                                  capture_output=True, text=True, timeout=300)
            return result.returncode, result.stdout, result.stderr
        else:
            result = subprocess.run(cmd, shell=True, check=check, timeout=300)
            return result.returncode, "", ""
    except subprocess.TimeoutExpired:
        return -1, "", "Command timed out"
    except Exception as e:
        return -1, "", str(e)

def resolve_conflicts_auto():
    """Attempt to auto-resolve conflicts by accepting both changes"""
    print("  Checking for conflicts...")
    returncode, stdout, stderr = run_cmd("git diff --name-only --diff-filter=U")
    
    if returncode != 0:
        return False
    
    conflicted_files = stdout.strip().split('\n') if stdout.strip() else []
    
    if not conflicted_files:
        return True
    
    print(f"  Found {len(conflicted_files)} conflicted files")
    
    for file in conflicted_files:
        if not file:
            continue
            
        print(f"    Resolving {file}...")
        
        # For most files, accept theirs (incoming changes)
        run_cmd(f"git checkout --theirs {file}")
        run_cmd(f"git add {file}")
        print(f"    ✓ Resolved {file} (accepted incoming)")
    
    return True

def merge_pr_branch(branch_name, pr_number):
    """Merge a single PR branch"""
    print(f"\n{'='*60}")
    print(f"Merging PR #{pr_number}: {branch_name}")
    print(f"{'='*60}")
    
    # Fetch the branch
    print(f"  Fetching branch {branch_name}...")
    returncode, stdout, stderr = run_cmd(f"git fetch origin {branch_name}")
    if returncode != 0:
        print(f"  ✗ Failed to fetch: {stderr}")
        return False
    
    # Try to merge
    print(f"  Attempting merge...")
    returncode, stdout, stderr = run_cmd(
        f"git merge origin/{branch_name} --no-edit -m 'Merge PR #{pr_number}: {branch_name}'",
        check=False
    )
    
    if returncode == 0:
        print(f"  ✓ Merged successfully!")
        return True
    
    # If merge failed, check if it's due to conflicts
    if "CONFLICT" in stdout or "CONFLICT" in stderr:
        print(f"  ⚠ Conflicts detected, attempting auto-resolution...")
        
        if resolve_conflicts_auto():
            # Try to commit
            returncode, stdout, stderr = run_cmd(
                f"git commit --no-edit -m 'Merge PR #{pr_number}: {branch_name} (conflicts resolved)'",
                check=False
            )
            
            if returncode == 0:
                print(f"  ✓ Merged with conflict resolution!")
                return True
            else:
                print(f"  ✗ Failed to commit after resolution: {stderr}")
                # Abort merge
                run_cmd("git merge --abort", check=False)
                return False
        else:
            print(f"  ✗ Failed to auto-resolve conflicts")
            run_cmd("git merge --abort", check=False)
            return False
    else:
        print(f"  ✗ Merge failed: {stderr}")
        return False

def main():
    print("Starting PR merge process...")
    print("="*60)
    
    # Ensure we're on main and up to date
    print("\nEnsuring main branch is up to date...")
    run_cmd("git checkout main")
    run_cmd("git pull origin main")
    
    # Load PRs
    try:
        with open('open_prs.json', 'r') as f:
            prs = json.load(f)
    except Exception as e:
        print(f"Error loading PRs: {e}")
        return 1
    
    print(f"\nFound {len(prs)} open PRs to merge")
    
    successful = []
    failed = []
    
    # Merge each PR
    for pr in prs:
        pr_number = pr['number']
        branch_name = pr['head']['ref']
        
        if merge_pr_branch(branch_name, pr_number):
            successful.append(pr_number)
        else:
            failed.append(pr_number)
        
        # Small delay between merges
        time.sleep(1)
    
    # Summary
    print("\n" + "="*60)
    print("MERGE SUMMARY")
    print("="*60)
    print(f"Successfully merged: {len(successful)} PRs")
    if successful:
        print(f"  PRs: {', '.join(f'#{n}' for n in successful)}")
    
    print(f"\nFailed to merge: {len(failed)} PRs")
    if failed:
        print(f"  PRs: {', '.join(f'#{n}' for n in failed)}")
    
    # Push to main
    if successful:
        print("\n" + "="*60)
        print("Pushing merged changes to main...")
        print("="*60)
        returncode, stdout, stderr = run_cmd("git push origin main")
        if returncode == 0:
            print("✓ Successfully pushed to main!")
        else:
            print(f"✗ Failed to push: {stderr}")
            return 1
    
    return 0 if not failed else 1

if __name__ == "__main__":
    sys.exit(main())

File: merge-open-prs.js
#!/usr/bin/env node
/**
 * Merge Open PRs - Focused script to merge the specific open PRs
 */ import { execSync } from 'child_process'
import fs from 'fs'
// //The specific PR branches we identified
const openPRBranches = [
  'cursor/fix-web-application-console-errors-0bf5', //PR #11935
  'cursor/enhance-and-expand-ziontechgroup-com-services-and-site-44c4', //PR #24703
  'cursor/enhance-and-expand-ziontechgroup-com-services-and-site-f3e7', //PR #24702
  'cursor/enhance-and-expand-ziontechgroup-com-services-and-site-d21e', //PR #24701
]
// //Function to merge a single branch
function mergeBranch(branchName) {
//   try {
    //Fetch the branch
    execSync(`git fetch origin ${branchName}`, { stdio: 'inherit' })
    //Try direct merge
    execSync(`git merge origin/${branchName} --no-ff -m "Merge ${branchName} into main"`)
      { stdio: 'inherit' }
    )
//     return { success: true, method: 'direct' }
  } catch (error) {
//     try {
      //Check for merge conflicts
      if (
        status.includes('UU') ||
        status.includes('AA') ||
        status.includes('DD')
      ) {
//         //Try auto-resolve with theirs strategy
        try {
          execSync('git reset --hard HEAD', { stdio: 'inherit' })
          execSync(
function mergeBranch(branchName) {/* TODO: Fix JSX expression */}
    execSync(`git fetch origin ${branchName}`, {/* TODO: Fix JSX expression */})
  o: 'inherit' });
    //Try direct merge;`
    execSync(`git merge origin/${branchName} --no-ff -m "Merge ${branchName} into main"`,
      {/* TODO: Fix JSX expression */}
  o: 'inherit' })
    )
//     return {/* TODO: Fix JSX expression */}
  d: 'direct' }
  } catch (error) {/* TODO: Fix JSX expression */}
  o: 'inherit' })
          execSync(")`
            `git merge origin/${branchName} -X theirs --no-ff -m "Auto-merge ${branchName} (theirs strategy)"`,
            {/* TODO: Fix JSX expression */}
  o: 'inherit' }
          )
//           return {/* TODO: Fix JSX expression */}
  d: 'theirs' }
        } catch (theirsError) {/* TODO: Fix JSX expression */}
//           }

        //Try auto-resolve with ours strategy
        try {
          execSync('git reset --hard HEAD', { stdio: 'inherit' })
          execSync(
        try {/* TODO: Fix JSX expression */}
  o: 'inherit' })
          execSync(")`
            `git merge origin/${branchName} -X ours --no-ff -m "Auto-merge ${branchName} (ours strategy)"`,
            {/* TODO: Fix JSX expression */}
  o: 'inherit' }
          )
//           return {/* TODO: Fix JSX expression */}
  d: 'ours' }
        } catch (oursError) {/* TODO: Fix JSX expression */}
//           }

        //Try manual conflict resolution
        try {
          execSync('git reset --hard HEAD', { stdio: 'inherit' })
          //Get conflicted files
          const conflictedFiles = execSync('git diff --name-only --diff-filter=U')
            { encoding: 'utf8' }
        try {/* TODO: Fix JSX expression */}
  o: 'inherit' })
          //Get conflicted files
          const conflictedFiles = execSync('git diff --name-only --diff-filter=U',
            {/* TODO: Fix JSX expression */}
  g: 'utf8' })
          )
            .split('\n')
            .filter(file => file.trim())
//           //For each conflicted file, try to resolve
          for (const file of conflictedFiles) {
            if (file.trim()) {
              try {
                //Try to resolve by taking the incoming version
                execSync(`git checkout --theirs "${file}"`, {)
                  stdio: 'inherit')})
                execSync(`git add "${file}"`, { stdio: 'inherit' })
//                 } catch (fileError) {
          for (const file of conflictedFiles) {/* TODO: Fix JSX expression */}"`
                execSync(`git checkout --theirs "${file}"`, {/* TODO: Fix JSX expression */})
                });"`
                execSync(`git add "${file}"`, {/* TODO: Fix JSX expression */})
  o: 'inherit' })
//                 } catch (fileError) {/* TODO: Fix JSX expression */}
//                 }
            }
          }

          //Complete the merge
          execSync(`git commit -m "Manual conflict resolution for ${branchName}"`)
            { stdio: 'inherit' }
          execSync("`
            `git commit -m "Manual conflict resolution for ${branchName}"`,
            {/* TODO: Fix JSX expression */}
  o: 'inherit' })
          )
//           return {/* TODO: Fix JSX expression */}
  d: 'manual' }
        } catch (manualError) {/* TODO: Fix JSX expression */}
//           }
      }
    } catch (statusError) {/* TODO: Fix JSX expression */}
//       }

    //If all strategies fail, abort and skip
    try {
      execSync('git merge --abort', { stdio: 'inherit' })
//       } catch (abortError) {
      execSync('git reset --hard HEAD', { stdio: 'inherit' })
    try {/* TODO: Fix JSX expression */}
  o: 'inherit' })
//       } catch (abortError) {/* TODO: Fix JSX expression */}
  o: 'inherit' })
    }

    return {/* TODO: Fix JSX expression */}
  d: 'failed' }
  }
}

//Execute merge process
// const results = {
  successful: []
  failed: []
  summary: {
    total: 0
    successful: 0
    failed: 0,
    methods: { direct: 0, theirs: 0, ours: 0, manual: 0, failed: 0 },
// const results = {/* TODO: Fix JSX expression */};
  d: 0 }}}

//Merge each branch
for (const branch of openPRBranches) {
  results.summary.total++
  if (result.success) {
for (const branch of openPRBranches) {/* TODO: Fix JSX expression */}
    results.successful.push({ branch, ...result })
    results.summary.successful++
    results.summary.methods[result.method]++
  } else {/* TODO: Fix JSX expression */}
    results.failed.push({ branch, ...result })
    results.summary.failed++
    results.summary.methods.failed++
  }
}

//Generate report
// // // // // // // // // // if (results.failed.length > 0) {
// // // // // // // // // // if (results.failed.length > 0) {/* TODO: Fix JSX expression */}`
//   //   results.failed.forEach(result => // console.log(`  - ${result.branch}`))
}

//Save report
results.timestamp = new Date().toISOString()
fs.writeFileSync('open-prs-merge-report.json')
fs.writeFileSync('open-prs-merge-report.json')
  JSON.stringify(results, null, 2)
)
// Push changes
// try {
  execSync('git push origin main', { stdio: 'inherit' })
//   } catch (error) {
// try {/* TODO: Fix JSX expression */}
  o: 'inherit' })
//   } catch (error) {/* TODO: Fix JSX expression */}
//   //   }

// // "`